metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/datahub_listener.py

```diff
@@ -0,0 +1,411 @@
+            iolet.urn for iolet in get_task_outlets(task) if isinstance(iolet, _Entity)
+        )
+
+        # Write the lineage to the datajob object.
+        datajob.inlets.extend(input_urns)
+        datajob.outlets.extend(output_urns)
+        datajob.fine_grained_lineages.extend(fine_grained_lineages)
+
+        # Merge in extra stuff that was present in the DataJob we constructed
+        # at the start of the task.
+        if complete:
+            original_datajob = self._datajob_holder.get(str(datajob.urn), None)
+        else:
+            self._datajob_holder[str(datajob.urn)] = datajob
+            original_datajob = None
+
+        if original_datajob:
+            logger.debug("Merging start datajob into finish datajob")
+            datajob.inlets.extend(original_datajob.inlets)
+            datajob.outlets.extend(original_datajob.outlets)
+            datajob.fine_grained_lineages.extend(original_datajob.fine_grained_lineages)
+
+            for k, v in original_datajob.properties.items():
+                if k not in datajob.properties:
+                    datajob.properties[k] = v
+
+        # TODO: Deduplicate inlets/outlets.
+
+        # Write all other OL facets as DataHub properties.
+        if task_metadata:
+            for k, v in task_metadata.job_facets.items():
+                datajob.properties[f"openlineage_job_facet_{k}"] = Serde.to_json(
+                    redact_with_exclusions(v)
+                )
+
+            for k, v in task_metadata.run_facets.items():
+                datajob.properties[f"openlineage_run_facet_{k}"] = Serde.to_json(
+                    redact_with_exclusions(v)
+                )
+
+    @hookimpl
+    @run_in_thread
+    def on_task_instance_running(
+        self,
+        previous_state,
+        task_instance: "TaskInstance",
+        session: "Session",  # This will always be QUEUED
+    ):
+        self._set_log_level()
+
+        # This if statement mirrors the logic in https://github.com/OpenLineage/OpenLineage/pull/508.
+        if not hasattr(task_instance, "task"):
+            logger.warning(
+                f"No task set for task_id: {task_instance.task_id} - "
+                f"dag_id: {task_instance.dag_id} - run_id {task_instance.run_id}"
+            )
+            return
+
+        logger.debug(
+            f"DataHub listener got notification about task instance start for {task_instance.task_id}"
+        )
+
+        # Render templates in a copy of the task instance.
+        # This is necessary to get the correct operator args in the extractors.
+        task_instance = copy.deepcopy(task_instance)
+        task_instance.render_templates()
+
+        dagrun: "DagRun" = task_instance.dag_run
+        task = task_instance.task
+        dag = task.dag
+
+        self._task_holder.set_task(task_instance)
+
+        # Handle async operators in Airflow 2.3 by skipping deferred state.
+        # Inspired by https://github.com/OpenLineage/OpenLineage/pull/1601.
```

Comment:
```suggestion
        # Inspired by https://github.com/OpenLineage/OpenLineage/pull/1601
```
Period messed up the url for me